Fix inconsistent previous attributes

ricardograca opened the pull request:
* Related Issues: #1520, #326

## Introduction

This makes `previousAttributes()` and `previous()` behave like users expect them to, and always return the actual previous attributes.

Also updates the documentation to make it clearer.

## Motivation

The current behavior causes some confusion, which isn't helped by the documentation not being very clear about it, and it seems to be a bit different from what other ORMs do. It's also not very useful to have the *previous* attributes be reset back to match the *current* attributes on each `save()` or `destroy()` call. This makes these methods far less useful since they can only be used to check for the previous attributes before the model is saved or destroyed.

## Proposed solution

This changes the `previous()` and `previousAttributes()` methods so that their values are not reset back to the current attributes when saving or destroying a model. This means that the previous attributes are accessible after any of these operations, making them way more useful.

This is a breaking change, but currently it's a bit useless having these methods return the same values that are already set on the model after a `save()`, so I doubt it will cause problems for anyone.

Fixes #1520.

## Alternatives considered

Implementing additional methods to get the previous attributes before a `save()`, but that would require adding more properties to models, more methods, and more chance for confusion. It would mean this change wouldn't break backwards compatibility though.

ricardograca commented:
@ErisDS @kirrg001 Care to check this out? This makes your custom `_updatedAttributes` implementation redundant, but it also changes the behavior of `previousAttributes()` so you may want to test these changes beforehand.

kirrg001 commented:
@ricardograca Sorry for delay - i was on vacation 😎Will look at it asap.

kirrg001 commented:
Is [this line](https://github.com/bookshelf/bookshelf/blob/0.13.3/src/model.js#L1094) still needed? 🤔 Because we no longer reset the previous attributes in the `reset` fn. Just curious.

ricardograca commented:
Good question. Probably not. Definitive answer in a few minutes.

ricardograca commented:
@kirrg001 Doesn't seem to be needed. On top of that `options.previousAttributes` is not documented anywhere, so I doubt many people are using it. For anyone using it it's just a matter of changing `options.previousAttributes` to `model.previousAttributes()` in the event listeners.

I'll wait for further reviews before changing this, so I can do it all in one go.

kirrg001 commented:
If i fetch relations with `withRelated`, `_previousAttributes` is `{}`.

ricardograca commented:
That doesn't seem right. I also just found one other somewhat related test that would pass in such a situation when it should fail. There's also no test for that particular scenario. I'll have to fix that.

ricardograca commented:
@kirrg001 All reported issues fixed, and I added some tests to check for them.

kirrg001 commented:
@ricardograca Great to hear - i'll look at them asap :)

kirrg001 commented:
> If i fetch relations with withRelated, the relational models have _previousAttributes: {}.

This is resolved 👍 

---

But when i install this PR as tarball, i am having a different problem. Maybe unrelated to this PR.

> onFetchingCollection ('fetching:collection')

This event no longer passes the `options` - it's always undefined.

First argument: is correct, the collection
Second argument: a function (?)
Third argument: undefined (which should be options)

ricardograca commented:
That hasn't changed AFAIK. Looking at the relevant code, the `options` are being provided to `triggerThen` as the third argument, so it must be something before that. I'll investigate, but it's probably not related to these changes.

kirrg001 commented:
@ricardograca If you have a guess, let me know. I've also debugged a little, suddenly the `options` and `columns` values change after the `fetching` event.

ricardograca commented:
Found the problem... it's something really simple.

kirrg001 commented:
@ricardograca great 👍 Let me know what it was :)

ricardograca commented:
Check out #1857.

kirrg001 commented:
Thanks! I'll look again at both on Monday 🙃

kirrg001 commented:
Sorry for late feedback.

If i forge a new model e.g. `models.Post.forge(data)`, why is the `previousAttributes` set to the `attributes`? 🤔This was different before and i wonder if we should change it.

kirrg001 commented:
Example: If you have previously used `model.hasChanged(key)` in a generic context, it was

- true if the model was new
- true if the key of the model got updated (model is not new)

ricardograca commented:
Hmm, it's been a while since I've worked on this but I'm almost certain that I did it like that because it's how it was already behaving and I didn't want to break backwards compatibility, otherwise I agree that having the `previousAttributes` be `{}` for a new model instance would make more sense.

There's also no test for `hasChanged()` without fetch which sucks... :(

I'll take a look at it and see if that was indeed the case or if there was any other reason.

Thank you for your comments.

ricardograca commented:
Hey @kirrg001 sorry for the long delay. I've looked at the changes and you are correct about `previousAttributes()` returning `attributes`. It should return `{}` when forging a new model indeed.

However, after a fetch, save or even eager loading it will return the current attributes. This is the correct behavior according to the docs and it's how it was already done before.

I've added more tests to check for the correct behavior and fixed the tests related with `previousAttributes()` on new forged models.

kirrg001 commented:
@ricardograca Will look at this asap. I am currently a little time constrained.

ricardograca commented:
@kirrg001 Care to take a look?

kirrg001 commented:
@ricardograca Yes, on my list for this week 👍

kirrg001 commented:
@ricardograca Could you pls rebase? 👍

ricardograca commented:
@kirrg001 Done.

kirrg001 commented:
Let's say i fetch a user and it returns

```
{attributes: {status: active, ...more attributes...}, previousAttributes: {}}
```

Now i call `user.set('status', 'active')` and `user.save()` (status did not change !)

I am listening on the user updated event and i receive:

```
{attributes: {status: active}, previousAttributes: {}, changed: {}}
```

Is that expected? I am wondering why the previous attributes object is empty in the updated event 🤔 I think as soon as you save a model without fields being modified, i am receiving `_previousAttributes: {}`, which feels inconsistent? Let me know :)

ricardograca commented:
@kirrg001 Are you sure you're testing with the latest changes and the correct branch? [This test](https://github.com/bookshelf/bookshelf/pull/1848/files#diff-080138e620928f31b725eed6a4b2ec9fR1580) ensures that calling `previousAttributes()` will return the current attributes if no changes were done to the model after fetch.

I haven't checked the case with the `updated` event yet though.

ricardograca commented:
I've added a few more tests to ensure the behavior you mention is observed: https://github.com/bookshelf/bookshelf/pull/1848/commits/3cdfa6e797400b23145c74829b58cf75fb77b56d

kirrg001 commented:
I am on the correct commit. I just double checked.

Steps:

- fetch
- user status is active
- user.set('status', 'active')
- user.save()
- updated event -> prev attrs are empty

ricardograca commented:
As you can see the test case is passing, and it seems to do exactly what you're saying. Maybe you have some plugin interfering with this, or some other event listener, or something else.

I'm writing a simple stand-alone test case that you can use to try to reproduce the behavior, but at this point I think Bookshelf itself is working fine. Oh yeah, I remember you had a custom property in Ghost models for storing the proper previous attributes, before this change. Could it be related to that?

ricardograca commented:
Here is the gist: https://gist.github.com/ricardograca/da68dc091fcbe15c395558c166fd7200

kirrg001 commented:
> Oh yeah, I remember you had a custom property in Ghost models for storing the proper previous attributes, before this change. Could it be related to that?

I have commented out all the logic for that.

> Here is the gist: https://gist.github.com/ricardograca/da68dc091fcbe15c395558c166fd7200

Thanks will test asap!!

kirrg001 commented:
The gist works for me too. It must be something else. I am looking into it now.

kirrg001 commented:
I figured it out. That was quick 🤣

https://gist.github.com/kirrg001/b8650a544db01541828d50fbcd1a227d

Bookshelf returns an empty `_previousAttributes` object if you fetch on a collection.
And if you then operate on a model of this collection, you end up that all further operations e.g. event hooks contain an empty previous attributes object.

ricardograca commented:
Nice work! I'll fix it this weekend hopefully. Apart from that are there any other issues?

ricardograca commented:
@kirrg001 Fixed. Nice catch there :1st_place_medal:
